java/org/apache/catalina/webresources/AbstractArchiveResourceSet.java

```diff
@@ -211,6 +216,43 @@ public final WebResource getResource(String path) {
         String webAppMount = getWebAppMount();
         WebResourceRoot root = getRoot();
 
+
+        /*
+         * This initializes (when necessary) and checks the jarContents, which
+         * is a highly efficient index of the files stored in the jar. If
+         * jarContents reports that this resource definitely does not contain
+         * the path, we can end this method and move on to the next jar.
+         *
+         * Note: the initialization is thread-safe because multiple simultaneous
+         * threads will create a complete and valid copy, then set the shared
+         * pointer. This guarantees the shared pointer will always go to a
+         * valid object. The cost of multiple copies is small since only one of
+         * them will be long-lived.
+         */
+        try {
+            if ((root.getContext() != null) && (root.getContext().getParent()) != null &&
+                    (((Host) root.getContext().getParent()).getFastClasspathScanning())) {
+
+                if (jarContents == null ||
+                        ((System.currentTimeMillis() - prevJarOpenTime > jarRefreshTime) &&
+                            !openJarFile().equals(jar))) {
+                    try {
+                        jar = openJarFile();
```

Comment:
You call `openJarFile()` twice here. This will increase `org.apache.catalina.webresources.AbstractArchiveResourceSet#archiveUseCount` twice and later it won't be GC-ed by `org.apache.catalina.webresources.AbstractArchiveResourceSet#gc`.
I think you need to call `org.apache.catalina.webresources.AbstractArchiveResourceSet#closeJarFile` after creating the JarContents

Reply:
Resolved with latest commit